src/mytardis_client/endpoints.py

```diff
@@ -1,196 +1,104 @@
 """Information about MyTardis endpoints"""
 
-from typing import Literal, Optional
+import re
+from typing import Any, Literal, get_args
+from urllib.parse import urlparse
 
-from pydantic import BaseModel
-
-from src.mytardis_client.objects import MyTardisObject
+from pydantic import RootModel, field_serializer, field_validator
 
 MyTardisEndpoint = Literal[
+    "/datafileparameter",
+    "/datafileparameterset",
     "/dataset",
+    "/datasetparameter",
     "/datasetparameterset",
     "/dataset_file",
     "/experiment",
+    "/experimentparameter",
     "/experimentparameterset",
     "/facility",
     "/group",
     "/institution",
     "/instrument",
     "/introspection",
+    "/parametername",
     "/project",
+    "/projectparameter",
     "/projectparameterset",
+    "/replica",
     "/schema",
     "/storagebox",
     "/user",
 ]
 
+_MYTARDIS_ENDPOINTS = list(get_args(MyTardisEndpoint))
+
+
+def list_mytardis_endpoints() -> list[str]:
+    """List the names of all MyTardis endpoints"""
+    return _MYTARDIS_ENDPOINTS
+
+
+uri_regex = re.compile(r"^/api/v1/([a-z_]{1,})/[0-9]{1,}/$")
+
+
+def validate_uri(value: Any) -> str:
+    """Validator for a URI string to ensure that it matches the expected form of a URI"""
+    if not isinstance(value, str):
+        raise TypeError(f'Unexpected type for URI: "{type(value)}"')
+    endpoint = uri_regex.search(value.lower())
+    if not endpoint:
+        raise ValueError(
+            f'Passed string value "{value}" is not a well formatted MyTardis URI'
+        )
+    endpoint_str = endpoint.group(1)
+
+    candidate_endpoint = f"/{endpoint_str.lower()}"
+    if candidate_endpoint not in _MYTARDIS_ENDPOINTS:
+        raise ValueError(f'Unknown endpoint: "{endpoint_str}"')
+    return value
+
+
+def resource_uri_to_id(uri: str) -> int:
+    """Gets the id from a resource URI
+
+    Takes resource URI like: http://example.org/api/v1/experiment/998
+    and returns just the id value (998).
+
+    Args:
+        uri: str - the URI from MyTardis
+
+    Returns:
+        The integer id that maps to the URI
+    """
+    uri_sep: str = "/"
+    return int(urlparse(uri).path.rstrip(uri_sep).split(uri_sep).pop())
```

Comment:
Curious why you chose pop() over slicing? Either works so this is not a criticism, just curious :)

Reply:
Good question. I actually just moved this function from a different source location - I didn't change the body of it, so it was doing the same thing previously. You may be right that slice is a better choice - there's probably no need to bother mutating the temporary `list` by removing the last element, when it disappears after this scope ends anyway.

I'd be happy with that change, but my inclination is to avoid changing it in this PR, since it's somewhat orthogonal to the purpose

Reply:
Agreed :)